### ISSUE TITLE
Unapplied plugins are only available to subprojects

### ISSUE THREAD

**Issue by cesar1000** [gradle/gradle#12878]
### Expected Behavior
I have a binary plugin that bundles a number of Kotlin DSL scripts, and I'm applying them to my build like this:

```kotlin
// /build.gradle.kts
plugins {
    id("com.twitter.gradle-plugin") version "0.0.10"
    id("com.twitter.publish") version "0.0.10"
}
```

Since all plugins are bundled together, I'd like to use `apply false` at the root to remove the need to specify the version number repeatedly. I added a new `root-configuration` script that's meant to be applied to the root project and looks like this:

```kotlin
// <my plugin>/root.configuration.gradle.kts

package com.twitter

plugins {
    // Make all other plugins available without the need for explicit versioning.
    id("com.twitter.kotlin-library-config") apply false
    id("com.twitter.functional-tests") apply false
    id("com.twitter.publish") apply false
    id("com.twitter.gradle-plugin") apply false
}
```

My expectation is that this would allow me to apply my plugins like this in the root project:

```kotlin
// /build.gradle.kts

plugins {
    id("com.twitter.root-configuration") version "0.0.10"
    id("com.twitter.gradle-plugin")
    id("com.twitter.publish")
}
```

### Current Behavior
In the current behavior, the root project fails to find the plugin:

```
* What went wrong:
Plugin [id: 'com.twitter.gradle-plugin'] was not found in any of the following sources:

- Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
- Plugin Repositories (plugin dependency must include a version number for this source)
```

_However_, the the version-less declaration does work in subprojects:

```kotlin
// /test/build.gradle.kts

plugins {
    id("com.twitter.gradle-plugin") // it works!
}
```

### Context
The [documentation](https://docs.gradle.org/current/userguide/plugins.html#sec:subprojects_plugins_dsl) does state explicitly that `apply false` enables the use of the version-less declaration in subprojects. I think that my example for the use of this mechanism in the root project itself is perfectly reasonable and, from the point of view of usage, the restriction feels arbitrary.

### Your Environment
I verified this behavior in 6.4-rc-1.

**Comment by vlsi** [comment]
@cesar1000 , would you please list the full reproducer? It is hard to guess your setup based on what samples you provide so far :-/

**Comment by cesar1000** [comment]
Sure: https://github.com/cesar1000/gradle-subproject-plugins. I hope that makes sense!

**Comment by vlsi** [comment]
Great.

`./gradlew buildEnvironment` might help to figure out the cause in similar cases.

```
plugins {
        id("com.twitter.test.root-configuration") version "0.0.1"
        id("com.twitter.test.publish") version "0.0.1" // this works
        //id("com.twitter.test.publish") // but this doesn't
        `java-library`
}
```

produces

```
> Task :buildEnvironment

------------------------------------------------------------
Root project
------------------------------------------------------------

classpath
+--- com.twitter.test.root-configuration:com.twitter.test.root-configuration.gradle.plugin:0.0.1
|    \--- com.twitter.gradle.test:gradle-configuration-plugin:0.0.1
\--- com.twitter.test.publish:com.twitter.test.publish.gradle.plugin:0.0.1
     \--- com.twitter.gradle.test:gradle-configuration-plugin:0.0.1
```

What that means is `id("com.twitter.test.publish") version "0.0.1"` results in Gradle trying to download `com.twitter.test.publish:com.twitter.test.publish.gradle.plugin:0.0.1` artifact (note it has machine-generated artifactId).

That is why if you omit the version, it won't really know which version of `com.twitter.test.publish.gradle.plugin` artifact do you mean.

This is similar: https://github.com/gradle/gradle/issues/10990

Unfortunately, the problem seems to be more involved than a mere absence of `com.twitter.test.publish.gradle.plugin` on the classpath.

Solution
--------

So what should work is a **settings** plugin that declares the relevant versions into the plugin management.
Have you tried it?

In other words, add the following to the `settings.gradle.kts`

```
buildscript {
    dependencies {
        classpath("com.twitter.test.root-configuration:root-configuration:0.0.1")
    }
}

apply(plugin = "com.twitter.root-configuration")
```

The plugin would be able to configure `pluginManagement` and/or apply project plugins as you need.

**Comment by cesar1000** [comment]
Thanks for your answer, @vlsi! I wanted to apply the plugin in the root project rather than a settings file since I'm configuring my repositories in an init script  – which I understand is somewhat of an odd requirement :)

However, my main point with this issue is that the inconsistency _feels_ like a bug. On the surface, both `plugins` blocks look the same, but their behavior is different. I understand that this is what's going on under the covers: the first stage of the compilation of the root project runs under a bootstrap class loader (which doesn't contain any plugins at the time), while the first stage of the compilation of the subproject is running under the root project class loader (which contains the plugin already). I think that it would be valuable for the documentation to specify these behaviors to facilitate interpreting the results.

Also, as it is, I'm forced to replicate the version number, even if all plugins are technically already available in the classpath. I think that my intent is reasonable: register all plugins when the root plugin is applied so that subsequent applications don't need to replicate the version number (or just trust Gradle to find them in the classpath even if they haven't been applied).
